examples/dcutr/src/main.rs

```diff
@@ -155,21 +122,30 @@ fn main() -> Result<(), Box<dyn Error>> {
         }
     }
 
-    let behaviour = Behaviour {
-        relay_client: client,
-        ping: ping::Behaviour::new(ping::Config::new()),
-        identify: identify::Behaviour::new(identify::Config::new(
-            "/TODO/0.0.1".to_string(),
-            local_key.public(),
-        )),
-        dcutr: dcutr::Behaviour::new(local_peer_id),
-    };
+    let mut swarm = libp2p::builder::SwarmBuilder::new()
+        .with_existing_identity(generate_ed25519(opts.secret_key_seed))
+        .with_async_std()
+        .with_custom_tcp(tcp::Config::default().port_reuse(true).nodelay(true))
+        .with_relay()
+        .with_other_transport(|keypair| {
+            quic::async_std::Transport::new(quic::Config::new(&keypair))
+                .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
+        })
+        .no_more_other_transports()
+        .with_dns()
+        .await
+        .with_behaviour(|keypair, relay_behaviour| Behaviour {
+            relay_client: relay_behaviour,
+            ping: ping::Behaviour::new(ping::Config::new()),
+            identify: identify::Behaviour::new(identify::Config::new(
+                "/TODO/0.0.1".to_string(),
+                keypair.public(),
+            )),
+            dcutr: dcutr::Behaviour::new(keypair.public().to_peer_id()),
+        })
+        .build();
```

Comment:
Do we need `.build` or can that not be implied by `with_behaviour`? Perhaps `.build` should take the closure? I think it is worth optimizing that users don't need to call "unnecessary" functions here.

Reply:
I am undecided. I like a `Builder` having a final `build` method. Including `build` in `with_behaviour` does not make it explicit that it is the final step. Including `with_behaviour` in `build` does not make it explicit that it wants a `NetworkBehaviour`.

I am fine with as is. In case you feel strongly about it, can you think of any other method names?

Reply:
> Including `with_behaviour` in `build` does not make it explicit that it wants a `NetworkBehaviour`.

It would be explicit because it takes an argument. Looking at it just as a builder-pattern, I think it is okay to include all required parameter in the `build` function. Traditionally, you would require them in the builder constructor but we cannot do that because the behaviour may need the identity key etc. So moving that to the `build` function is the next best thing IMO.

Reply:
`with_existing_identity` takes a `Keypair`, `with_tcp_config` takes a `TcpConfig`, `with_other_transport` takes another `Transport`. I find `build` taking a `NetworkBehaviour` inconsistent.

Reply:
Would `build_with_behaviour` be too verbose?

Reply:
In my eyes it combines two unrelated actions. `with_behaviour` is like `with_tcp` and `with_other_transport` in regards to providing some input. `build` is a final step combining all of the previous inputs. It does not make sense to me to combine `build` with one of the input actions (i.e. `with_behaviour`).